### PR TITLE
initial support for @NaturalIds

### PR DESCRIPTION
**Pull request by gavinking**
I need to do some more thinking about this because `@SafeVarargs` doesn't work on interfaces so we get the stupid stupid useless fucking unchecked array creation warning. Dumbest thing in Java.

### CODE REVIEW

**Comment by gavinking** [comment]
@DavideD any idea why these CI runs are failing? Everything working fine locally.

**Comment by DavideD** [comment]
Yes, it has to do with the latest ORM 5.5 SNAPSHOT: https://github.com/hibernate/hibernate-reactive/pull/643/checks?check_run_id=2024823063

**Comment by DavideD** [comment]
You will see the error if you run this command (I think):
```
./gradlew build -Pdb=DB2 -Pdocker -PhibernateOrmVersion='[5.5,5.6)' -PskipOrmVersionParsing -PenableJBossSnapshotsRep
```

**Comment by gavinking** [comment]
OK, so I guess this doesn't really affect us doing a release that targets 5.4, right?

**Comment by DavideD** [comment]
I don't think so, but we will continue to see errors until it's compatible

**Comment by DavideD** [comment]
By the way, the problem is due to this commit in ORM: https://github.com/hibernate/hibernate-orm/commit/ed3bbf15e476b064b6a55a80ac0c1570a4287bf0

**Comment by DavideD** [comment]
I think i'm going to disable the build for 5.5 for now.
Unless somebody finds a solution that doesn't require a release of 5.4

**Comment by gavinking** [comment]
Thanks @DavideD

File: hibernate-reactive-core/src/test/java/org/hibernate/reactive/NaturalIdTest.java [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+        thing2.naturalKey = "abc123";
+        thing2.version = 2;
+        test( context,
+                getSessionFactory()
+                        .withSession( session -> session.persist(thing1, thing2).thenCompose( v -> session.flush() ) )
+                        .thenCompose( v -> getSessionFactory().withSession(
+                                session -> session.find( Thing.class, composite(
+                                        id("naturalKey", "abc123"),
+                                        id("version", 1)
+                                ) )
```
Comment (gavinking) [inline hibernate-reactive-core/src/test/java/org/hibernate/reactive/NaturalIdTest.java:44]: Alright, I've now got an API for this which isn't *terrible*, though it's also far from perfect. (It's better if you use the typesafe JPA metamodel.)

I propose including this in the release, and I've clearly marked it `@Incubating`.

Comment (gavinking) [reply hibernate-reactive-core/src/test/java/org/hibernate/reactive/NaturalIdTest.java:44]: @DavideD? @Sanne?

Comment (DavideD) [reply hibernate-reactive-core/src/test/java/org/hibernate/reactive/NaturalIdTest.java:44]: LGTM

Comment (DavideD) [reply hibernate-reactive-core/src/test/java/org/hibernate/reactive/NaturalIdTest.java:44]: I guess an alternative would be a DSL:
```
session.find( Thing.class, id("naturalKey", "abc123")
                                 .and("version", 1) )
```

but I don't think I like this solution more

Comment (gavinking) [reply hibernate-reactive-core/src/test/java/org/hibernate/reactive/NaturalIdTest.java:44]: I don't much like it either, even though it's perhaps a little more ergonomic.

**Comment by gavinking** [comment]
I'm going to merge it. It's marked experimental, and I would rather have it there so people can experiment with it.

**Pull request merged**